docker/DOCKER.md

```diff
@@ -0,0 +1,167 @@
+# Docker Images for balena CLI
+
+Docker images with balena CLI and docker-in-docker.
+
+## Available architectures
+
+- `rpi`
+- `armv7hf`
+- `aarch64` (debian only)
+- `amd64`
+- `i386`
+
+## Basic Usage
+
+Here's a small example of running a single, detached container
+in the background and using `docker exec` to run balena CLI commands.
+
+```
+$ docker run --detach --privileged --network host --name cli --rm -it balenalib/amd64-debian-balenacli /bin/bash
```

Comment:
`balenalib/amd64-debian-balenacli`

iirc, we had discussed ([flowdock thread](https://www.flowdock.com/app/rulemotion/resin-projects/threads/enZ9VZOK1lFNUhq5N2AfTssafFb)) that this PR could be merged before the images were actually available in Docker Hub. I may have said that I was OK with this, but I am having second thoughts. :-) What was the benefit of merging this PR before (well before?) transformers / something else was available to actually use the Dockerfiles and upload the images to Docker Hub?  What I am specifically concerned with is that users will see this documentation and Dockerfile in the CLI repo, will try to use it, and then will create CLI issues ("it doesn't work!"), will contact support, and will be disappointed... 

The PR could be left around unmerged, until the time comes. Or, if helpful, we could merge the Dockerfiles without this DOCKER.md file. Or we could replace the current contents of DOCKER.md with:

```
Note: The files in this folder are experimental and a work in progress, and not
ready for end users.
```
But if it's experimental and work in progress, then why merge it...
Ah! Note that transformers may also be triggered (if we implement it so) when commits are pushed to a branch like this one. Well, at least build tests are triggered by merely pushing commits, so we could probably find a way of making sure that whatever we needed was triggered to.

Anyway, perhaps it's just a case of me not seeing (or having already forgotten) "the reason why". :-)

Reply:
@pdcastro you probably aren't forgetting our reason to merge these early, cause I don't think I provided one!

I personally wanted these Dockerfiles put somewhere more "stable" than the playground. As part of my fear of losing track of good bits of code, I like to know everything is stashed in the correct repository with eyes on it so it doesn't get lost or forgotten. That doesn't require by any means that this needs to be merged right now!

I also don't know how long it will be until transformers are ready to pick up this kind of project (I likely don't have the bandwidth to be much help getting them there), and I'd hate to look at these files next year and forget how the heck they work!

We could totally leave this PR as a draft, or merged without the DOCKER.md, or with a disclaimer at the top. Leaving it as a draft is probably easiest?

Reply:
Sorry for the delay in replying. This excellent bit of code will not be lost if left here as a draft PR... But also, someone at some point will have to pick up the effort of using transformers to build the CLI images and upload them to Dockerhub, otherwise the excellent bits of code will just stay here for a long time. :-)

Hey, what about modifying `DOCKER.md` to include instructions of how users could manually build the images on their own machines / devices? Using Docker on the laptops, or balenaEngine on balenaOS devices. E.g., including `docker build` and `balena-engine build` command lines in addition to `docker run` command lines, and removing references to images that don't yet exist, like `balenalib/amd64-debian-balenacli`. Then this could be merged now and start being of some use to end users. The following bit of the installation instructions could then be amended as well:

> https://github.com/balena-io/balena-cli/blob/master/INSTALL-ADVANCED.md#npm-installation
> _This [sample Dockerfile](https://gist.github.com/pdcastro/5d4d96652181e7da685a32caf629dd44) shows the CLI installation steps on an Ubuntu 18.04 base image._

Instead of that sample Dockerfile, the advanced installation instructions would refer to these Dockerfiles. Perhaps a new section could be added to `INSTALL-ADVANCED.md`, with a short introduction to the concept of running the CLI through docker containers, and a link to DOCKER.md for the command lines and examples.

By the way, here's a half baked idea / question. I know for a fact that some users insist on running the CLI on devices (e.g. because they find it easier to write a shell script around some CLI commands, instead of programming with the Node SDK or making direct use of balena's HTTP API). But an issue for them is that we don't have standalone installers for ARM, which would allow the CLI to be used on the Raspberry Pi. Sometimes they get around it by installing the CLI with npm, but it starts getting messy as it needs the CLI dependencies for compilation of native modules etc. With these Dockerfiles, there'd be the obvious solution of users copying-and-pasting the Dockerfile contents to their application's Dockerfile, or the other way around. But - and here's the half baked question - could they add the CLI as a service to their microservices / multicontainer application? E.g. adding the CLI Dockerfile as a service in the `docker-compose.yml` of their application. In this case, could they do the equivalent of `docker exec -it` inside an app container, to exec the CLI in the other container? Perhaps using DBUS or some other simpler / clever way of doing it? Actually, if the balenaEngine socket is exposed in the container, they could perhaps include a docker client in the app container and literally run `docker exec -it` - maybe we could document some shell `alias` command or shell function to make it user friendly to run `balena` (the CLI) across containers. Could input and output files be placed in a shared volume, for CLI commands that operate on files? ... Anyway, just a thought. :-)

Reply:
I can definitely adjust the readme and documentation to include instructions of how users could manually build the images on their own machines / devices! If you'll recall, this readme initially had **Build** instructions near the top since that was the only way to get an image until we pipelined everything. I can move it back to the top and elaborate a bit more, since I use this as a template for a few projects as it is!

Regarding half-baked, I definitely think there's something here as far as running a balena-cli service in existing apps, I had the same idea. The only thing I haven't solved for yet is the best way to communicate between the containers as you mentioned. We have a similar problem in testbot/exclave where we want to execute commands in a parallel testbot container (ideally without an API or REST interface).

You may be on to something with the docker client though. If we could provide instructions or a pre-built client that would allow executing commands in another container as long as the host balena socket is exposed, that would be applicable to many uses cases, not just the cli.

Maybe two components with pre-packaged files and examples?
1. `balena-exec` unpack this binary your app (preinstalled in balenalib base images) to execute commands in other services
2. `balena-alias` with a balena-cli service (depends on balena-exec) use this alias to run `balena` commands as if the cli were installed (using shared volumes, etc)

Reply:
> I definitely think there's something here as far as running a balena-cli service in existing apps, I had the same idea.

There is an additional consideration though, before we went ahead with something like that. Currently, we don't advise users to run the CLI on devices, in fact we occasionally advise them not to, e.g.:
https://github.com/balena-io/balena-cli/issues/1450#issuecomment-535024573

The CLI uses Amplitude.com for command usage tracking and Sentry.io for error reporting with the assumption that users run the CLI on their dev workstations, which is a low-volume scenario. Even so, we occasionally hit the Sentry.io rate limit and monthly limit. If users were to deploy the CLI in fleets of thousands of devices, such usage tracking and error reporting would be degraded. So I think there would need to be a production discussion and code and docs adjustments before we did that. Technically, we could try some sort of auto detection to disable Amplitude/Sentry, and/or we could document flags or env vars to disable those features. (Requests to Amplitude use the balena API as a proxy, so work could be done there too.)

> If you'll recall, this readme initially had Build instructions near the top since that was the only way to get an image until we pipelined everything. I can move it back to the top and elaborate a bit more, since I use this as a template for a few projects as it is!

Yes it would be good, but now with the view that the build step is required rather than optional, because "official" images don't exist in `balenalib` or Docker Hub (so no references to `balenalib/amd64-debian-balenacli`. The point above about running the CLI in automated environments may still apply to some extent though. Why would users want to build CLI images and run the CLI using Docker? If it's to run the CLI interactively in their dev workstation then fine, but I suspect it would most often be related to running in CI environments or on devices. That product discussion may be unavoidable... Do we want to make it easier (or even necessary) to disable Amplitude and Sentry.io?...

Reply:
JF brainstorm topic (restricted access): https://jel.ly.fish/faccdba2-becc-40be-8f74-f34b9703c5d5